sklearn/neighbors/_nearest_centroid.py

```diff
@@ -255,7 +312,95 @@ def predict(self, X):
         """
         check_is_fitted(self)
 
-        X = self._validate_data(X, accept_sparse="csr", reset=False)
-        return self.classes_[
-            pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
-        ]
+        if len(np.unique(self.class_priors_, return_counts=True)[0]) == 1:
+            X = self._validate_data(X, accept_sparse="csr", reset=False)
+            return self.classes_[
+                pairwise_distances_argmin(X, self.centroids_, metric=self.metric)
+            ]
+        else:
+            d = self._decision_function(X)
+            return self.classes_.take(d.argmax(1))
+
+    def _decision_function(self, X):
+        check_is_fitted(self, "centroids_")
+
+        X = self._validate_data(X, reset=False, accept_sparse="csr")
+
+        # check if this works with sparse matrix formats later
+        discriminant_score = np.empty(
+            (X.shape[0], self.classes_.size), dtype=np.float64
+        )
+
+        for cur_class in range(self.classes_.size):
```

Comment:
I think that we can make use of `pairwise_distance` here and just standardize the data and the centroids with the pooled std. dev. before to compute the distance.

```python
        X = self._validate_data(X, copy=True, reset=False, accept_sparse="csr")

        # check if this works with sparse matrix formats later
        discriminant_score = np.empty(
            (X.shape[0], self.classes_.size), dtype=np.float64
        )

        # Standardize X and the centroids based on within-class std_dev
        X /= self.within_class_std_
        centroids = self.centroids_ / self.within_class_std_

        for class_idx in range(self.classes_.size):
            discriminant_score[:, class_idx] = - pairwise_distances(
                X, centroids[[class_idx], :], metric=self.metric, n_jobs=self.n_jobs,
            ).ravel()
            if self.metric == "euclidean":
                discriminant_score[:, class_idx] **= 2
            discriminant_score[:, class_idx] += 2 * np.log(
                self.class_priors_[class_idx]
            )
```

Something that now is not clear to me is that we allow for the manhattan distance. Then, I assume that we can pass the parameter `self.metric` in pairwise distance to adapt the computation of the discriminant score. For instance the above code would provide something like this:

<img width="530" alt="image" src="https://github.com/scikit-learn/scikit-learn/assets/7454015/425c42a7-b361-4aa7-b3c0-2264f24e9d77">

Also above I introduce `n_jobs`: we should check if it could be beneficial in practice or not.

Reply:
Manhattan distance is just another name for L<sup>1</sup>-norm and while there are research papers on LDA based on different norms (e.g. [1](https://proceedings.mlr.press/v196/nagananda22a/nagananda22a.pdf) and [2](https://ranger.uta.edu/~chqding/papers/R1PCA.pdf)) I don't think they are quite standardized yet so I'm actually thinking of leaving the __decision_function_ and _pred_proba_ and _log_pred_proba_ undefined for distance metrics other than Euclidean.

And out of curiosity, I looked up, the [Laplace distribution](https://en.wikipedia.org/wiki/Laplace_distribution) may be a good choice as the L1-norm version of the normal distribution for calculating the posterior probabilities (equation from ESLII):
![image](https://github.com/scikit-learn/scikit-learn/assets/6528822/6d6f32d2-d479-4beb-a4c6-49d548e0f0cc)

But again, I think we should leave the __decision_function_ and _pred_proba_ and _log_pred_proba_ undefined for distance metrics other than Euclidean. Thoughts?

For `n_jobs`, it's highly dependent on cases but I see that it's an option in `LogisticRegression` for one-vs-rest classification, so we can at least give users an option.

Reply:
To clarify, we can still use `Manhattan` metric to predict, but not to estimate posterior probability or to model decision function.

Reply:
This is fine by me to raise an error or make the methods available only in the euclidean case.